File: TheWatch-iOS/ViewModels/Logging/LogViewerViewModel.swift
/**
 * ┌──────────────────────────────────────────────────────────────────────┐
 * │ WRITE-AHEAD LOG                                                      │
 * ├──────────────────────────────────────────────────────────────────────┤
 * │ File:         LogViewerViewModel.swift (ViewModels/Logging)          │
 * │ Purpose:      @Observable ViewModel for the production Log Viewer.   │
 * │               Subscribes to loggingPort.logPublisher for real-time   │
 * │               streaming. Manages filter/search state, expandable     │
 * │               entry IDs, and share-sheet export of filtered output.  │
 * │ Created:      2026-03-24                                             │
 * │ Author:       Claude                                                 │
 * │ Dependencies: Combine, LoggingPort, LogEntry, LogLevel              │
 * │                                                                      │
 * │ Usage example:                                                       │
 * │   @State private var viewModel = LogViewerViewModel()                │
 * │   // In onAppear: viewModel.configure(loggingPort: WatchLogger.shared.port!) │
 * │                                                                      │
 * │ Architecture:                                                        │
 * │ - All data flows through LoggingPort (hexagonal port/adapter)        │
 * │ - No direct SwiftData/Core Data access from this ViewModel           │
 * │ - logPublisher gives a Combine AnyPublisher<LogEntry, Never>         │
 * │ - query() used for initial load and pull-to-refresh                  │
 * │ - Export writes to temp directory, returns URL for UIActivityVC      │
 * │                                                                      │
 * │ Filter semantics:                                                    │
 * │ - Level filter: shows entries AT or ABOVE selected level             │
 * │ - Search: case-insensitive on messageTemplate + sourceContext        │
 * │ - Both compose with AND logic                                        │
 * │                                                                      │
 * │ Expandable entries:                                                  │
 * │ - Tap row to toggle; expanded shows properties, correlationId, etc  │
 * │ - Multiple simultaneous expansions supported                         │
 * │                                                                      │
 * │ Export format (plain text):                                          │
 * │   [2026-03-24T10:35:00Z] [information] [LocationCoordinator]        │
 * │   Location updated to 30.2672,-97.7431 in HIGH_ACCURACY mode        │
 * │   Properties: {Lat: 30.2672, Lng: -97.7431}                         │
 * │   CorrelationId: sos-abc123                                          │
 * │   ---                                                                │
 * └──────────────────────────────────────────────────────────────────────┘
 */

import Foundation
import Combine
import SwiftUI

@Observable
final class LogViewerViewModel {

    // MARK: - Published State

    var entries: [LogEntry] = []
    var filteredEntries: [LogEntry] = []
    var selectedLevel: LogLevel? = nil
    var searchQuery: String = ""
    var expandedEntryIds: Set<String> = []
    var isRefreshing: Bool = false
    var isLiveStreaming: Bool = true
    var exportFileURL: URL? = nil

    // MARK: - Computed

    var entryCount: Int { entries.count }
    var filteredCount: Int { filteredEntries.count }

    func countForLevel(_ level: LogLevel) -> Int {
        entries.count { $0.level == level }
    }

    // MARK: - Private

    private var loggingPort: (any LoggingPort)?
    private var cancellables = Set<AnyCancellable>()

    private let dateFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "HH:mm:ss.SSS"
        return f
    }()

    private let isoFormatter: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()

    // MARK: - Configuration

    /// Call this once to wire the ViewModel to the LoggingPort.
    /// Typically called from .onAppear in the View.
    func configure(loggingPort: any LoggingPort) {
        guard self.loggingPort == nil else { return } // Already configured
        self.loggingPort = loggingPort
        initialLoad()
        startLiveStream()
    }

    // MARK: - Initial Load

    private func initialLoad() {
        guard let port = loggingPort else { return }
        Task { @MainActor in
            let loaded = await port.query(limit: 500, minLevel: .verbose, sourceContext: nil)
            self.entries = loaded
            self.applyFilters()
        }
    }

    // MARK: - Live Streaming

    private func startLiveStream() {
        guard let port = loggingPort else { return }
        port.logPublisher
            .receive(on: DispatchQueue.main)
            .sink { [weak self] entry in
                guard let self else { return }
                self.entries.insert(entry, at: 0)
                // Cap at 2000 in memory
                if self.entries.count > 2000 {
                    self.entries = Array(self.entries.prefix(2000))
                }
                self.applyFilters()
            }
            .store(in: &cancellables)
    }

    // MARK: - Pull-to-Refresh

    func refresh() {
        guard let port = loggingPort else { return }
        isRefreshing = true
        Task { @MainActor in
            let loaded = await port.query(limit: 500, minLevel: .verbose, sourceContext: nil)
            self.entries = loaded
            self.isRefreshing = false
            self.applyFilters()
        }
    }

    // MARK: - Filters

    func setLevelFilter(_ level: LogLevel?) {
        selectedLevel = level
        applyFilters()
    }

    func setSearchQuery(_ query: String) {
        searchQuery = query
        applyFilters()
    }

    private func applyFilters() {
        var result = entries

        if let minLevel = selectedLevel {
            result = result.filter { $0.level.rawValue >= minLevel.rawValue }
        }

        if !searchQuery.isEmpty {
            result = result.filter {
                $0.messageTemplate.localizedCaseInsensitiveContains(searchQuery) ||
                $0.sourceContext.localizedCaseInsensitiveContains(searchQuery) ||
                $0.renderedMessage.localizedCaseInsensitiveContains(searchQuery)
            }
        }

        filteredEntries = result
    }

    // MARK: - Expand / Collapse

    func toggleExpanded(_ entryId: String) {
        if expandedEntryIds.contains(entryId) {
            expandedEntryIds.remove(entryId)
        } else {
            expandedEntryIds.insert(entryId)
        }
    }

    func isExpanded(_ entryId: String) -> Bool {
        expandedEntryIds.contains(entryId)
    }

    // MARK: - Clear

    func clearLogs() {
        entries = []
        filteredEntries = []
        expandedEntryIds = []
    }

    // MARK: - Formatting

    func formatTime(_ date: Date) -> String {
        dateFormatter.string(from: date)
    }

    // MARK: - Export to Share

    /// Writes filtered entries to a temp .txt file and returns the URL.
    /// Use this URL with UIActivityViewController / ShareLink.
    func exportToFile() -> URL? {
        guard !filteredEntries.isEmpty else { return nil }

        var text = ""
        text += "TheWatch Log Export — \(isoFormatter.string(from: Date()))\n"
        text += "Filter: level=\(selectedLevel.map { "\($0)" } ?? "ALL"), query=\"\(searchQuery)\"\n"
        text += "Entries: \(filteredEntries.count)\n"
        text += String(repeating: "═", count: 72) + "\n\n"

        for entry in filteredEntries {
            text += "[\(isoFormatter.string(from: entry.timestamp))] [\(entry.level)] [\(entry.sourceContext)]\n"
            text += entry.renderedMessage + "\n"
            if !entry.properties.isEmpty {
                let props = entry.properties.map { "  \($0.key): \($0.value)" }.joined(separator: "\n")
                text += "  Properties:\n\(props)\n"
            }
            if let cid = entry.correlationId {
                text += "  CorrelationId: \(cid)\n"
            }
            if let ex = entry.exception {
                text += "  Exception: \(ex)\n"
            }
            text += "---\n"
        }

        let formatter = DateFormatter()
        formatter.dateFormat = "yyyyMMdd_HHmmss"
        let timestamp = formatter.string(from: Date())
        let fileName = "thewatch_logs_\(timestamp).txt"
        let tempURL = FileManager.default.temporaryDirectory.appendingPathComponent(fileName)

        do {
            try text.write(to: tempURL, atomically: true, encoding: .utf8)
            exportFileURL = tempURL
            return tempURL
        } catch {
            return nil
        }
    }
}

File: TheWatch-Android/app/src/main/java/com/thewatch/app/ui/logging/LogViewerViewModel.kt
/**
 * ┌──────────────────────────────────────────────────────────────────────┐
 * │ WRITE-AHEAD LOG                                                      │
 * ├──────────────────────────────────────────────────────────────────────┤
 * │ File:         LogViewerViewModel.kt                                  │
 * │ Purpose:      Hilt ViewModel for the production Log Viewer screen.   │
 * │               Collects from LoggingPort.observe() for real-time      │
 * │               streaming and LoggingPort.query() for pull-to-refresh. │
 * │               Manages filter/search state, expandable entry IDs,     │
 * │               and export-to-file for sharing filtered log output.    │
 * │ Created:      2026-03-24                                             │
 * │ Author:       Claude                                                 │
 * │ Dependencies: Hilt, LoggingPort, LogEntry, LogLevel, Coroutines     │
 * │                                                                      │
 * │ Usage example:                                                       │
 * │   @Composable                                                        │
 * │   fun LogViewerRoute(navController: NavController) {                 │
 * │       val vm: LogViewerViewModel = hiltViewModel()                   │
 * │       LogViewerScreen(navController, vm)                             │
 * │   }                                                                  │
 * │                                                                      │
 * │ Architecture:                                                        │
 * │ - All data flows through LoggingPort (hexagonal port/adapter)        │
 * │ - No direct Room/Firestore access from this ViewModel               │
 * │ - observe() gives a hot Flow<LogEntry> for live streaming            │
 * │ - query() is used for initial load and pull-to-refresh               │
 * │ - Export writes to app cache dir, returns shareable URI              │
 * │                                                                      │
 * │ Filter semantics:                                                    │
 * │ - Level filter: shows entries AT or ABOVE selected level             │
 * │ - Search: case-insensitive match on messageTemplate + sourceContext  │
 * │ - Both filters compose (AND logic)                                   │
 * │                                                                      │
 * │ Expandable entries:                                                  │
 * │ - Tap a log row to toggle expansion                                  │
 * │ - Expanded view shows: properties map, correlationId, exception      │
 * │ - Multiple entries can be expanded simultaneously                    │
 * │                                                                      │
 * │ Export format (plain text):                                          │
 * │   [2026-03-24T10:35:00.123Z] [INFO] [LocationCoordinator]           │
 * │   Location updated to 30.2672,-97.7431 in HIGH_ACCURACY mode        │
 * │   Properties: {Lat=30.2672, Lng=-97.7431, Mode=HIGH_ACCURACY}       │
 * │   CorrelationId: sos-abc123                                          │
 * │   ---                                                                │
 * └──────────────────────────────────────────────────────────────────────┘
 */
package com.thewatch.app.ui.logging

import android.content.Context
import android.content.Intent
import android.net.Uri
import androidx.core.content.FileProvider
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.thewatch.app.data.logging.LogEntry
import com.thewatch.app.data.logging.LogLevel
import com.thewatch.app.data.logging.LoggingPort
import dagger.hilt.android.lifecycle.HiltViewModel
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.io.File
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import javax.inject.Inject

/**
 * UI state for the Log Viewer screen.
 *
 * @property entries All log entries collected so far (newest first)
 * @property filteredEntries Entries after level + search filters applied
 * @property selectedLevel Minimum severity filter (null = show all)
 * @property searchQuery Free-text search against messageTemplate and sourceContext
 * @property expandedEntryIds Set of entry IDs currently expanded in the UI
 * @property isRefreshing True while a pull-to-refresh query is in progress
 * @property isLiveStreaming True while the observe() collector is active
 * @property entryCount Total count of unfiltered entries
 * @property exportUri URI of the last exported file (null if no export yet)
 */
data class LogViewerUiState(
    val entries: List<LogEntry> = emptyList(),
    val filteredEntries: List<LogEntry> = emptyList(),
    val selectedLevel: LogLevel? = null,
    val searchQuery: String = "",
    val expandedEntryIds: Set<String> = emptySet(),
    val isRefreshing: Boolean = false,
    val isLiveStreaming: Boolean = true,
    val entryCount: Int = 0,
    val exportUri: Uri? = null
)

@HiltViewModel
class LogViewerViewModel @Inject constructor(
    private val loggingPort: LoggingPort,
    @ApplicationContext private val appContext: Context
) : ViewModel() {

    private val _uiState = MutableStateFlow(LogViewerUiState())
    val uiState: StateFlow<LogViewerUiState> = _uiState.asStateFlow()

    private val timeFormatter: DateTimeFormatter = DateTimeFormatter
        .ofPattern("HH:mm:ss.SSS")
        .withZone(ZoneId.systemDefault())

    private val isoFormatter: DateTimeFormatter = DateTimeFormatter.ISO_INSTANT

    init {
        initialLoad()
        startLiveStream()
    }

    // ── Initial Load ──────────────────────────────────────────────

    private fun initialLoad() {
        viewModelScope.launch {
            val entries = loggingPort.query(limit = 500)
            _uiState.value = _uiState.value.copy(
                entries = entries,
                entryCount = entries.size
            )
            applyFilters()
        }
    }

    // ── Live Streaming ────────────────────────────────────────────

    private fun startLiveStream() {
        viewModelScope.launch {
            loggingPort.observe(LogLevel.Verbose).collect { entry ->
                val current = _uiState.value
                val updated = listOf(entry) + current.entries
                // Cap at 2000 entries in memory to prevent OOM
                val capped = if (updated.size > 2000) updated.take(2000) else updated
                _uiState.value = current.copy(
                    entries = capped,
                    entryCount = capped.size
                )
                applyFilters()
            }
        }
    }

    // ── Pull-to-Refresh ───────────────────────────────────────────

    fun refresh() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isRefreshing = true)
            val entries = loggingPort.query(limit = 500)
            _uiState.value = _uiState.value.copy(
                entries = entries,
                entryCount = entries.size,
                isRefreshing = false
            )
            applyFilters()
        }
    }

    // ── Filters ───────────────────────────────────────────────────

    fun setLevelFilter(level: LogLevel?) {
        _uiState.value = _uiState.value.copy(selectedLevel = level)
        applyFilters()
    }

    fun setSearchQuery(query: String) {
        _uiState.value = _uiState.value.copy(searchQuery = query)
        applyFilters()
    }

    private fun applyFilters() {
        var result = _uiState.value.entries

        _uiState.value.selectedLevel?.let { minLevel ->
            result = result.filter { it.level.ordinal >= minLevel.ordinal }
        }

        val query = _uiState.value.searchQuery
        if (query.isNotBlank()) {
            result = result.filter {
                it.messageTemplate.contains(query, ignoreCase = true) ||
                    it.sourceContext.contains(query, ignoreCase = true) ||
                    it.renderedMessage().contains(query, ignoreCase = true)
            }
        }

        _uiState.value = _uiState.value.copy(filteredEntries = result)
    }

    // ── Expand / Collapse ─────────────────────────────────────────

    fun toggleExpanded(entryId: String) {
        val current = _uiState.value.expandedEntryIds
        _uiState.value = _uiState.value.copy(
            expandedEntryIds = if (entryId in current) current - entryId else current + entryId
        )
    }

    fun isExpanded(entryId: String): Boolean =
        entryId in _uiState.value.expandedEntryIds

    // ── Clear ─────────────────────────────────────────────────────

    fun clearLogs() {
        _uiState.value = _uiState.value.copy(
            entries = emptyList(),
            filteredEntries = emptyList(),
            expandedEntryIds = emptySet(),
            entryCount = 0
        )
    }

    // ── Export to Shareable File ──────────────────────────────────

    /**
     * Writes filtered entries to a plain-text file in the app cache directory
     * and returns a content:// URI suitable for sharing via Intent.ACTION_SEND.
     *
     * Export format per entry:
     *   [ISO-8601 timestamp] [LEVEL] [SourceContext]
     *   Rendered message text
     *   Properties: {key=value, ...}
     *   CorrelationId: <id>      (if present)
     *   Exception: <text>        (if present)
     *   ---
     */
    fun exportToFile(): Uri? {
        val entries = _uiState.value.filteredEntries
        if (entries.isEmpty()) return null

        val sb = StringBuilder()
        sb.appendLine("TheWatch Log Export — ${Instant.now()}")
        sb.appendLine("Filter: level=${_uiState.value.selectedLevel?.name ?: "ALL"}, query=\"${_uiState.value.searchQuery}\"")
        sb.appendLine("Entries: ${entries.size}")
        sb.appendLine("═".repeat(72))
        sb.appendLine()

        for (entry in entries) {
            sb.appendLine("[${isoFormatter.format(entry.timestamp)}] [${entry.level.name}] [${entry.sourceContext}]")
            sb.appendLine(entry.renderedMessage())
            if (entry.properties.isNotEmpty()) {
                sb.appendLine("  Properties: ${entry.properties}")
            }
            entry.correlationId?.let { sb.appendLine("  CorrelationId: $it") }
            entry.exception?.let { sb.appendLine("  Exception: $it") }
            sb.appendLine("---")
        }

        val exportDir = File(appContext.cacheDir, "log_exports")
        exportDir.mkdirs()
        val timestamp = DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss")
            .withZone(ZoneId.systemDefault())
            .format(Instant.now())
        val file = File(exportDir, "thewatch_logs_$timestamp.txt")
        file.writeText(sb.toString())

        val uri = FileProvider.getUriForFile(
            appContext,
            "${appContext.packageName}.fileprovider",
            file
        )
        _uiState.value = _uiState.value.copy(exportUri = uri)
        return uri
    }

    /**
     * Creates a share intent for the exported log file.
     * Call exportToFile() first, then pass the returned URI here.
     */
    fun createShareIntent(uri: Uri): Intent {
        return Intent(Intent.ACTION_SEND).apply {
            type = "text/plain"
            putExtra(Intent.EXTRA_STREAM, uri)
            putExtra(Intent.EXTRA_SUBJECT, "TheWatch Log Export")
            addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        }
    }

    // ── Formatting Helpers ────────────────────────────────────────

    fun formatTime(instant: Instant): String = timeFormatter.format(instant)

    fun levelColor(level: LogLevel): Long = when (level) {
        LogLevel.Verbose -> 0xFF808080
        LogLevel.Debug -> 0xFF4FC3F7
        LogLevel.Information -> 0xFF81C784
        LogLevel.Warning -> 0xFFFFB74D
        LogLevel.Error -> 0xFFE57373
        LogLevel.Fatal -> 0xFFFF1744
    }
}
